### PR TITLE
GitHub Actions: Upgrade to actions/checkout@v4

### PR DESCRIPTION
**Pull request by cclauss**
* https://github.com/actions/checkout/releases
* https://github.com/actions/setup-python/blob/main/docs/advanced-usage.md#allow-pre-releases

### CODE REVIEW

File: .github/workflows/test.yml [hunk 1]
```diff
@@ -6,15 +6,16 @@ jobs:
     strategy:
       fail-fast: false
       matrix:
-        python-version: [3.8, 3.9, '3.10', 3.11, '3.12-dev', pypy3.8, pypy3.9]
```
Comment (skirpichev) [inline .github/workflows/test.yml:9]: Don't drop, please pypy3.8.

**Review by skirpichev — CHANGES_REQUESTED** [review]

File: .github/workflows/test.yml (hunk 1 above)
Comment (cclauss) [reply .github/workflows/test.yml:9]: https://www.pypy.org/download.html

File: .github/workflows/test.yml [hunk 2]
```diff
@@ -6,15 +6,16 @@ jobs:
...
+        python-version: [3.8, 3.9, '3.10', 3.11, 3.12, pypy3.9, pypy3.10]
```
Comment (skirpichev) [inline .github/workflows/test.yml:9]:
```suggestion
        python-version: [3.8, 3.9, '3.10', 3.11, 3.12, pypy3.8, pypy3.9, pypy3.10]
```

File: .github/workflows/test.yml (hunk 1 above)
Comment (skirpichev) [reply .github/workflows/test.yml:9]: https://github.com/actions/setup-python/blob/main/docs/advanced-usage.md#available-versions-of-python-and-pypy

**Review by skirpichev — APPROVED** [review]

**Pull request merged**